modules/navigation/nav_utils.h

```diff
@@ -150,6 +150,34 @@ struct ClosestPointQueryResult {
 	RID owner;
 };
 
+enum PathfindingAlgorithm {
+	PATHFINDING_ALGORITHM_ASTAR = 0,
+};
+
+struct PathQueryParameters {
+	PathfindingAlgorithm pathfinding_algorithm = PATHFINDING_ALGORITHM_ASTAR;
+	RID map;
+	Vector3 start_position;
+	Vector3 target_position;
+	bool optimize = true;
+	uint32_t navigation_layers = 1;
+	bool calculate_path_length = false;
+	bool deferred_processing = false;
+	ObjectID callback_object_id = ObjectID();
+	StringName callback_method = StringName();
+};
+
+struct PathQueryResult {
+	Vector<Vector3> path;
+	real_t path_length = 0.0;
+};
+
+struct PathQueryTask {
+	PathQueryParameters query_parameters;
+	ObjectID result_object_id;
```

Comment:
What's the pros and cons of storing an object for the calling actor (`result_object_id`) _AND_ the receiving object in the `PathQueryParameters` (`callback_object_id`)? Maybe the callback object will be deleted if the actor is?

Reply:
Well here it seems like the crucial thing is the `callback_object_id` rather than the actor. If the `callback_object_id` is valid, you can go ahead, if not you can't. I don't think sending / storing / checking the actor ID is necessary if all the communication is via this object.

Reply:
Forget what I said about mentioned delete issues. Changed the query stuff from objects to RefCounted some weeks ago so it does not suffer the same problems as the agents had e.g. from an early user `free()` or `queue_free()` while still in the waiting room.

The `result_object_id` and `callback_object_id` are also split so callbacks can happen on objects that the users needs them, not fixated to the result, e.g. while the result is updated and used by the actor the callback that the result has been updated is received on another method or the method of an entirely different object e.g. a manager object that controls the ai actors.

The parameter `callback_object_id` is also optional in the parameters. If no `callback_object_id` or `callback_method` the query is processed immediately as was requested in devchat.

Reply:
removed from this pr